### PR TITLE
blockdev_inc_backup_pull_mode_vm_reboot: vm reboot test for pull

### PR DESCRIPTION
**Pull request by zhencliu**
Backup image in pull mode(sync: none), reboot vm while pulling
  data from exported fleecing image(exported by internal nbd)

Signed-off-by: Zhenchao Liu <zhencliu@redhat.com>

ID: 1909636

### CODE REVIEW

**Comment by zhencliu** [comment]
Aborted against qemu-kvm-5.1.0-14.module+el8.3.0+8438+644aff69.x86_64:
(1/1) type_specific.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot: FAIL: qemu aborted(core dumped) (134.07 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-12-22T20.46-4249662/results.html
JOB TIME   : 136.34 s

I don't have a qemu-kvm that includes the fix, so just remove the iothread to simulate the correct way qemu handles(the issue will not be triggered if iothread is not set):
(1/1) type_specific.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot: PASS (272.30 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-12-22T20.50-9d39e7d/results.html
JOB TIME   : 275.25 s

**Comment by zhencliu** [comment]
(3/3) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.x86_64.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot.q35: PASS (429.23 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-12-23T10.54-0cc00da/results.html
JOB TIME   : 3294.14 s

Run this case on a 'slow' node with virtio-scsi, which cost 3200+s, @ngu-niny , can we remove the virtio-scsi from the patch?

File: provider/backup_utils.py [hunk 1]
```diff
@@ -406,3 +409,52 @@ def format_storage_volume(img, filesystem, partition="mbr"):
...
+
+    map_cmd = '{qemu_img} map --output=json {args}'.format(
+        qemu_img=qemu_img, args=args)
+    result = process.run(map_cmd, ignore_status=False, shell=True)
+
+    for item in json.loads(result.stdout.decode().strip()):
+        if item['data'] is not state:
+            continue
+        io_cmd = '{io} -C -c "r {s} {l}" -f {fmt} {fn}'.format(
+            io=qemu_io, s=item['start'], l=item['length'],
```
Comment (ngu-niny) [inline provider/backup_utils.py:454]: @zhencliu Zhenchao, according to https://bugzilla.redhat.com/show_bug.cgi?id=1657095#c10, we need to divide the qemu-io length to be within 0 ~ 2147483136. So will you modify it? Or we can merge the patch first and then fix it together with the copyif() in other cases in a future patch.

Comment (zhencliu) [reply provider/backup_utils.py:459]: Nini, let me enhance it in this patch, will let you know once I finish it. Thanks.

Comment (zhencliu) [reply provider/backup_utils.py]: change done.

**Comment by zhencliu** [comment]
virtio-blk without iothread:
 (1/1) type_specific.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot: PASS (251.94 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-12-25T09.53-1041454/results.html
JOB TIME   : 254.46 s

virtio-blk + iothread (--iothread_scheme=rhv)
 (1/1) type_specific.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot: FAIL: qemu aborted(core dumped) (138.25 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kar/workspace/job-results/job-2020-12-25T10.00-315c897/results.html
JOB TIME   : 141.00 s

hi @ngu-niny , change done, can you please make a double check, thanks.

**Comment by ngu-niny** [comment]
LGTM. Test with data plane and virtio-blk enabled core dump while others PASS:

With data plane:
 (05/12) repeat1.Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot:  FAIL: qemu aborted(core dumped) (345.17 s)
 (07/12) repeat1.Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot:  PASS (492.79 s)
 (09/12) repeat2.Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot:  FAIL: qemu aborted(core dumped) (52.45 s)
(11/12) repeat2.Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot:  PASS (371.46 s)

Without data plane:
 (1/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot: PASS (395.57 s)
 (2/2) Host_RHEL.m8.u4.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.4.0.ppc64le.io-github-autotest-qemu.blockdev_inc_backup_pull_mode_vm_reboot: PASS (417.55 s)

Acked-by: Nini Gu <ngu@redhat.com>

**Pull request merged**